File: V2/DIDx509/Validation/PolicyValidators.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace DIDx509.Validation;

using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using DIDx509.Models;

/// <summary>
/// Validates the "subject" policy against a certificate chain.
/// </summary>
internal static class SubjectPolicyValidator
{
    [ExcludeFromCodeCoverage]
    internal static class ClassStrings
    {
        public static readonly string ErrorFailedToParsePolicyValue =
            string.Concat(DidX509Constants.ErrorInvalidSubjectPolicy, ": Failed to parse policy value");
        public static readonly string ErrorMustContainAtLeastOneAttribute =
            string.Concat(DidX509Constants.ErrorInvalidSubjectPolicy, ": Must contain at least one attribute");

        public const string ErrorRequiredAttributeNotFound =
            "Subject policy validation failed: Required attribute '{0}' not found in leaf certificate subject";
        public const string ErrorAttributeValueMismatch =
            "Subject policy validation failed: Attribute '{0}' value mismatch (expected '{1}', got '{2}')";
    }

    /// <summary>
    /// Validates that the leaf certificate subject contains all specified attributes.
    /// </summary>
    /// <param name="policy">The policy to validate.</param>
    /// <param name="chain">The certificate chain model.</param>
    /// <param name="errors">When this method returns, contains validation errors if validation failed.</param>
    /// <returns><see langword="true"/> if the policy validates; otherwise, <see langword="false"/>.</returns>
    public static bool Validate(DidX509Policy policy, CertificateChainModel chain, out List<string> errors)
    {
        errors = new List<string>();

        if (policy.ParsedValue is not Dictionary<string, string> requiredAttributes)
        {
            errors.Add(ClassStrings.ErrorFailedToParsePolicyValue);
            return false;
        }

        if (requiredAttributes.Count == 0)
        {
            errors.Add(ClassStrings.ErrorMustContainAtLeastOneAttribute);
            return false;
        }

        var leafSubject = chain.LeafCertificate.Subject;

        // Check if all required attributes are present in the leaf certificate subject
        foreach (var kvp in requiredAttributes)
        {
            string? actualValue = leafSubject.GetAttribute(kvp.Key);

            if (actualValue == null)
            {
                errors.Add(string.Format(ClassStrings.ErrorRequiredAttributeNotFound, kvp.Key));
                return false;
            }

            if (!string.Equals(actualValue, kvp.Value, System.StringComparison.Ordinal))
            {
                errors.Add(string.Format(ClassStrings.ErrorAttributeValueMismatch, kvp.Key, kvp.Value, actualValue));
                return false;
            }
        }

        return true;
    }
}

/// <summary>
/// Validates the "san" policy against a certificate chain.
/// </summary>
internal static class SanPolicyValidator
{
    [ExcludeFromCodeCoverage]
    internal static class ClassStrings
    {
        public static readonly string ErrorFailedToParsePolicyValue =
            string.Concat(DidX509Constants.ErrorInvalidSanPolicy, ": Failed to parse policy value");

        public const string ErrorNoSans =
            "SAN policy validation failed: Leaf certificate has no Subject Alternative Names";
        public const string ErrorRequiredSanNotFound =
            "SAN policy validation failed: Required SAN '{0}:{1}' not found in leaf certificate";
    }

    /// <summary>
    /// Validates that the leaf certificate contains the specified SAN.
    /// </summary>
    /// <param name="policy">The policy to validate.</param>
    /// <param name="chain">The certificate chain model.</param>
    /// <param name="errors">When this method returns, contains validation errors if validation failed.</param>
    /// <returns><see langword="true"/> if the policy validates; otherwise, <see langword="false"/>.</returns>
    public static bool Validate(DidX509Policy policy, CertificateChainModel chain, out List<string> errors)
    {
        errors = new List<string>();

        if (!(policy.ParsedValue is ValueTuple<string, string> sanTuple))
        {
            errors.Add(ClassStrings.ErrorFailedToParsePolicyValue);
            return false;
        }

        var leafExtensions = chain.LeafCertificate.Extensions;

        if (leafExtensions.San == null || leafExtensions.San.Count == 0)
        {
            errors.Add(ClassStrings.ErrorNoSans);
            return false;
        }

        // Check if the required SAN is present
        bool found = leafExtensions.San.Any(san =>
            string.Equals(san.Type, sanTuple.Item1, System.StringComparison.OrdinalIgnoreCase) &&
            san.ValueAsString != null &&
            string.Equals(san.ValueAsString, sanTuple.Item2, System.StringComparison.Ordinal));

        if (!found)
        {
            errors.Add(string.Format(ClassStrings.ErrorRequiredSanNotFound, sanTuple.Item1, sanTuple.Item2));
            return false;
        }

        return true;
    }
}

/// <summary>
/// Validates the "eku" policy against a certificate chain.
/// </summary>
internal static class EkuPolicyValidator
{
    [ExcludeFromCodeCoverage]
    internal static class ClassStrings
    {
        public static readonly string ErrorFailedToParsePolicyValue =
            string.Concat(DidX509Constants.ErrorInvalidEkuPolicy, ": Failed to parse policy value");

        public const string ErrorNoEku =
            "EKU policy validation failed: Leaf certificate has no Extended Key Usage extension";
        public const string ErrorRequiredEkuNotFound =
            "EKU policy validation failed: Required EKU OID '{0}' not found in leaf certificate";
    }

    /// <summary>
    /// Validates that the leaf certificate contains the specified EKU OID.
    /// </summary>
    /// <param name="policy">The policy to validate.</param>
    /// <param name="chain">The certificate chain model.</param>
    /// <param name="errors">When this method returns, contains validation errors if validation failed.</param>
    /// <returns><see langword="true"/> if the policy validates; otherwise, <see langword="false"/>.</returns>
    public static bool Validate(DidX509Policy policy, CertificateChainModel chain, out List<string> errors)
    {
        errors = new List<string>();

        if (policy.ParsedValue is not string requiredEkuOid)
        {
            errors.Add(ClassStrings.ErrorFailedToParsePolicyValue);
            return false;
        }

        var leafExtensions = chain.LeafCertificate.Extensions;

        if (leafExtensions.Eku == null || leafExtensions.Eku.Count == 0)
        {
            errors.Add(ClassStrings.ErrorNoEku);
            return false;
        }

        // Check if the required EKU OID is present
        bool found = leafExtensions.Eku.Any(oid =>
            string.Equals(oid, requiredEkuOid, System.StringComparison.Ordinal));

        if (!found)
        {
            errors.Add(string.Format(ClassStrings.ErrorRequiredEkuNotFound, requiredEkuOid));
            return false;
        }

        return true;
    }
}

/// <summary>
/// Validates the "fulcio-issuer" policy against a certificate chain.
/// </summary>
internal static class FulcioIssuerPolicyValidator
{
    [ExcludeFromCodeCoverage]
    internal static class ClassStrings
    {
        public static readonly string ErrorFailedToParsePolicyValue =
            string.Concat(DidX509Constants.ErrorInvalidFulcioPolicy, ": Failed to parse policy value");

        public const string ErrorNoFulcioIssuer =
            "Fulcio issuer policy validation failed: Leaf certificate has no Fulcio issuer extension";
        public const string ErrorExpectedGot =
            "Fulcio issuer policy validation failed: Expected '{0}', got '{1}'";
    }

    /// <summary>
    /// Validates that the leaf certificate contains the specified Fulcio issuer.
    /// </summary>
    /// <param name="policy">The policy to validate.</param>
    /// <param name="chain">The certificate chain model.</param>
    /// <param name="errors">When this method returns, contains validation errors if validation failed.</param>
    /// <returns><see langword="true"/> if the policy validates; otherwise, <see langword="false"/>.</returns>
    public static bool Validate(DidX509Policy policy, CertificateChainModel chain, out List<string> errors)
    {
        errors = new List<string>();

        if (policy.ParsedValue is not string requiredIssuerSuffix)
        {
            errors.Add(ClassStrings.ErrorFailedToParsePolicyValue);
            return false;
        }

        var leafExtensions = chain.LeafCertificate.Extensions;

        if (string.IsNullOrEmpty(leafExtensions.FulcioIssuer))
        {
            errors.Add(ClassStrings.ErrorNoFulcioIssuer);
            return false;
        }

        // The policy value should match the issuer URL without the "https://" prefix
        string expectedUrl = DidX509Constants.ProtocolHttps + requiredIssuerSuffix;

        if (!string.Equals(leafExtensions.FulcioIssuer, expectedUrl, System.StringComparison.Ordinal))
        {
            errors.Add(string.Format(ClassStrings.ErrorExpectedGot, expectedUrl, leafExtensions.FulcioIssuer));
            return false;
        }

        return true;
    }
}

File: native/rust/did/x509/src/policy_validators.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use crate::constants::*;
use crate::error::DidX509Error;
use crate::models::SanType;
use crate::san_parser;
use crate::x509_extensions;
use x509_parser::prelude::*;

/// Validate Extended Key Usage (EKU) policy
pub fn validate_eku(cert: &X509Certificate, expected_oids: &[String]) -> Result<(), DidX509Error> {
    let ekus = x509_extensions::extract_extended_key_usage(cert);

    if ekus.is_empty() {
        return Err(DidX509Error::PolicyValidationFailed(
            "EKU policy validation failed: Leaf certificate has no Extended Key Usage extension"
                .into(),
        ));
    }

    // Check that ALL expected OIDs are present
    for expected_oid in expected_oids {
        if !ekus.iter().any(|oid| oid == expected_oid) {
            return Err(DidX509Error::PolicyValidationFailed(format!(
                "EKU policy validation failed: Required EKU OID '{}' not found in leaf certificate",
                expected_oid
            )));
        }
    }

    Ok(())
}

/// Validate Subject Distinguished Name policy
pub fn validate_subject(
    cert: &X509Certificate,
    expected_attrs: &[(String, String)],
) -> Result<(), DidX509Error> {
    if expected_attrs.is_empty() {
        return Err(DidX509Error::PolicyValidationFailed(
            "Subject policy validation failed: Must contain at least one attribute".into(),
        ));
    }

    // Parse the certificate subject
    let subject = cert.subject();

    // Check that ALL expected attribute/value pairs match
    for (attr_label, expected_value) in expected_attrs {
        // Find the OID for this attribute label
        let oid = attribute_label_to_oid(attr_label).ok_or_else(|| {
            DidX509Error::PolicyValidationFailed(format!(
                "Subject policy validation failed: Unknown attribute '{}'",
                attr_label
            ))
        })?;

        // Find the attribute in the subject RDN sequence
        let mut found = false;
        let mut actual_value: Option<String> = None;

        for rdn in subject.iter() {
            for attr in rdn.iter() {
                if attr.attr_type().to_id_string() == oid {
                    found = true;
                    if let Ok(value) = attr.attr_value().as_str() {
                        actual_value = Some(value.to_string());
                        if value == expected_value {
                            // Exact match found, continue to next expected attribute
                            break;
                        }
                    }
                }
            }
            if found
                && actual_value
                    .as_ref()
                    .map(|v| v == expected_value)
                    .unwrap_or(false)
            {
                break;
            }
        }

        if !found {
            return Err(DidX509Error::PolicyValidationFailed(
                format!("Subject policy validation failed: Required attribute '{}' not found in leaf certificate subject", attr_label)
            ));
        }

        if let Some(actual) = actual_value {
            if actual != *expected_value {
                return Err(DidX509Error::PolicyValidationFailed(
                    format!("Subject policy validation failed: Attribute '{}' value mismatch (expected '{}', got '{}')", 
                        attr_label, expected_value, actual)
                ));
            }
        } else {
            return Err(DidX509Error::PolicyValidationFailed(format!(
                "Subject policy validation failed: Attribute '{}' value could not be parsed",
                attr_label
            )));
        }
    }

    Ok(())
}

/// Validate Subject Alternative Name (SAN) policy
pub fn validate_san(
    cert: &X509Certificate,
    san_type: &SanType,
    expected_value: &str,
) -> Result<(), DidX509Error> {
    let sans = san_parser::parse_sans_from_certificate(cert);

    if sans.is_empty() {
        return Err(DidX509Error::PolicyValidationFailed(
            "SAN policy validation failed: Leaf certificate has no Subject Alternative Names"
                .into(),
        ));
    }

    // Check that the expected SAN type+value exists
    let found = sans
        .iter()
        .any(|san| &san.san_type == san_type && san.value == expected_value);

    if !found {
        return Err(DidX509Error::PolicyValidationFailed(format!(
            "SAN policy validation failed: Required SAN '{}:{}' not found in leaf certificate",
            san_type.as_str(),
            expected_value
        )));
    }

    Ok(())
}

/// Validate Fulcio issuer policy
pub fn validate_fulcio_issuer(
    cert: &X509Certificate,
    expected_issuer: &str,
) -> Result<(), DidX509Error> {
    let fulcio_issuer = x509_extensions::extract_fulcio_issuer(cert);

    if fulcio_issuer.is_none() {
        return Err(DidX509Error::PolicyValidationFailed(
            "Fulcio issuer policy validation failed: Leaf certificate has no Fulcio issuer extension".into()
        ));
    }

    let actual_issuer = fulcio_issuer.unwrap();

    // The expected_issuer might not have the https:// prefix, so add it if needed
    let expected_url = if expected_issuer.starts_with("https://") {
        expected_issuer.to_string()
    } else {
        format!("https://{}", expected_issuer)
    };

    if actual_issuer != expected_url {
        return Err(DidX509Error::PolicyValidationFailed(format!(
            "Fulcio issuer policy validation failed: Expected '{}', got '{}'",
            expected_url, actual_issuer
        )));
    }

    Ok(())
}
